src/GSoft.Extensions.Mongo.ApplicationInsights/ApplicationInsightsEventSubscriber.cs

```diff
@@ -0,0 +1,159 @@
+    {
+        this._telemetryClient = telemetryClient;
+        this._options = options;
+        this._currentOperationsMap = new ConcurrentDictionary<int, ActivityAwareOperationHolder>();
+
+        this.Subscribe<CommandStartedEvent>(this.CommandStartedEventHandler);
+        this.Subscribe<CommandSucceededEvent>(this.CommandSucceededEventHandler);
+        this.Subscribe<CommandFailedEvent>(this.CommandFailedEventHandler);
+    }
+
+    private void CommandStartedEventHandler(CommandStartedEvent evt)
+    {
+        if (this._options.Telemetry.IgnoredCommandNames.Contains(evt.CommandName))
+        {
+            return;
+        }
+
+        var name = evt.GetCollectionName() is { } collectionName
+            ? $"{evt.DatabaseNamespace.DatabaseName}.{collectionName}.{evt.CommandName}"
+            : $"{evt.DatabaseNamespace.DatabaseName}.{evt.CommandName}";
+
+        var operation = this.StartActivityAwareDependencyOperation(name);
+
+        operation.Telemetry.Type = "mongodb";
+
+        operation.Telemetry.Target = evt.ConnectionId?.ServerId?.EndPoint switch
+        {
+            IPEndPoint endPoint => endPoint.Address + ":" + endPoint.Port,
+            DnsEndPoint endPoint => endPoint.Host + ":" + endPoint.Port,
+            _ => "unknown",
+        };
+
+        if (this._options.Telemetry.CaptureCommandText)
+        {
+            operation.Telemetry.Data = evt.Command.ToString();
+        }
+
+        // Originating activity must be captured AFTER that the operation is created
+        // Because ApplicationInsights SDK creates another intermediate Activity
+        var originatingActivity = Activity.Current;
+
+        this._currentOperationsMap.TryAdd(evt.RequestId, new ActivityAwareOperationHolder(operation, originatingActivity));
+    }
+
+    private void CommandSucceededEventHandler(CommandSucceededEvent evt)
+    {
+        if (!this._currentOperationsMap.TryRemove(evt.RequestId, out var operation))
+        {
+            return;
+        }
+
+        using (operation)
+        {
+            operation.Telemetry.Duration = evt.Duration;
+            operation.Telemetry.Success = true;
+        }
+    }
+
+    private void CommandFailedEventHandler(CommandFailedEvent evt)
+    {
+        if (!this._currentOperationsMap.TryRemove(evt.RequestId, out var operation))
+        {
+            return;
+        }
+
+        using (operation)
+        {
+            operation.Telemetry.Duration = evt.Duration;
+            operation.Telemetry.Success = false;
+
+            if (!operation.Telemetry.Properties.ContainsKey("Exception"))
+            {
+                operation.Telemetry.Properties.Add("Exception", evt.Failure.ToString());
+            }
+        }
+    }
+
+    private IOperationHolder<DependencyTelemetry> StartActivityAwareDependencyOperation(string name)
+    {
+        if (Activity.Current is { } activity && TracingHelper.IsMongoActivity(activity))
```

Comment:
I would invert the conditional check here to reduce the general nesting of this method and short-circuit.